CORE-6871 Consensual Finality flow first version

lankydan opened the pull request:
Added `ConsensualFinalityFlow` and `ConsensualReceiveFinalityFlow` which
are both accessed by `ConsensualLedgerService`.

Included `ledger-consensual-serialization` and
`ledger-common-serialization` as runtime dependencies of
`ledger-consensual-impl`.

Made a number of services `corda.system` services as needed.

lankydan committed:
CORE-6871 finality flow

CORE-6871 fix serialization test

CORE-6871 fix serialization test

CORE-6871 fix serialization test

CORE-6871 fix serialization test

CORE-6871 alpha

CORE-6871 Rebase

CORE-6871 Add TODO comments tickets

CORE-6871 Some tidy up

CORE-6871 Consensual flow cleanup

CORE-6871 Detekt + tidy up

CORE-6871 Finality flow fix after rebase

CORE-6871 Finality flow tidy + fix tests

CORE-6871 Finality flow first version

Added `ConsensualFinalityFlow` and `ConsensualReceiveFinalityFlow` which
are both accessed by `ConsensualLedgerService`.

Included `ledger-consensual-serialization` and
`ledger-common-serialization` as runtime dependencies of
`ledger-consensual-impl`.

Made a number of services `corda.system` services as needed.

lankydan commented:
This is built off of `vlajos/CORE-5114-signedtransaction-serializer-amqp` so I have diffed compared to that to get an idea of the changes.

vlajos commented on components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/sandbox/impl/FlowSandboxServiceImpl.kt:
```diff
@@ -83,6 +83,8 @@ class FlowSandboxServiceImpl @Activate constructor(
     private val log = loggerFor<FlowSandboxServiceImpl>()
 
+    // does this lookup based on the current sandbox? Doesn't feel like it does since its not given any sandbox information
+    // so surely there is no way that it can discover PROTOTYPE services
```
The current custom serializers are not PROTOTYPE services.

vlajos commented on libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/transaction/ConsensualTransactionBuilderImpl.kt:
```diff
@@ -18,109 +18,129 @@ import net.corda.v5.ledger.consensual.ConsensualState
+// We should be able to write AMQP serializers and discover them using OSGi, but I don't think they can rely on PROTOTYPE
+// services, which most of our services are (although they possibly don't need to be)?
+// Otherwise, we'd need to use the "currently executing fiber" to get access to these services, however that would
+// basically make the fiber's context the equivalent of servicehub.
```
In theory the signedTx serializers are ready in the vlajos/CORE-5114-signedtransaction-serializer-amqp branch.

Re PROTOTYPE question:
https://github.com/corda/corda-runtime-os/pull/2213/files#diff-e626587a875b7d0048665080f0e26a70150050781d232512580c8e351d458339R87
> the current custom serializers are not PROTOTYPE services.

Although I am not too sure about the whole context.

lankydan replied:
I'll remove these comments since I wrote these before I started using your serializers.

vlajos commented on libs/ledger/ledger-consensual-serialization/src/main/kotlin/net/corda/ledger/consensual/transaction/serialization/internal/ConsensualSignedTransactionSerializer.kt:
```diff
@@ -1,43 +1,47 @@
 
 import net.corda.ledger.consensual.impl.transaction.ConsensualSignedTransactionImpl
-import net.corda.ledger.consensual.transaction.serialization.ConsensualSignedTransactionImplContainer
 import net.corda.ledger.consensual.transaction.serialization.ConsensualSignedTransactionImplVersion
```
Probably this one could drop `Impl` like the others.

vlajos commented on libs/ledger/ledger-consensual-serialization/src/main/kotlin/net/corda/ledger/consensual/transaction/serialization/internal/ConsensualSignedTransactionSerializer.kt:
```diff
@@ -1,43 +1,47 @@
+        return ConsensualSignedTransactionProxy(
             ConsensualSignedTransactionImplVersion.VERSION_1,
-            obj.wireTransaction,
+            // we need an internal interface
```
What do you mean? (I am planning to tackle the remaining/ pending serialization related issues.)

lankydan replied:
I think having an internal interface to cast to instead of casting to the implementation would be tidier, has the overhead of an extra interface though.

blsemo commented on components/flow/flow-service/src/main/kotlin/net/corda/flow/external/events/impl/ExternalEventManagerImpl.kt:
```diff
@@ -78,10 +78,7 @@ class ExternalEventManagerImpl(
-            "Processing received external event response of type ${externalEventResponse.payload.javaClass.name} " +
-                    "with id $requestId"
-        }
+        log.debug { "Processing received external event response of type with id $requestId" }
```
That log message looks wrong after removing the type info.

blsemo commented on libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/ConsensualLedgerServiceImpl.kt:
```diff
@@ -1,41 +1,68 @@
+        sessions: List<FlowSession>
+    ): ConsensualSignedTransaction {
+        /*
+        Need [doPrivileged] due to [contextLogger] being used in the flow's constructor.
+        Creating the executing the SubFlow must be independent otherwise the security manager causes issues with Quasar.
```
Will that be causing an issue for subflows created by user code in CorDapps?

lankydan replied:
We ran into issues with this a while back for starting flows, however I have started subflows in tests that use `contextLogger` and they seemed to work.

blsemo commented on libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/flows/finality/ConsensualFinalityFlow.kt:
```diff
@@ -0,0 +1,137 @@
+            // Split send and receive since we have to use [FlowMessaging.sendAll] and [FlowMessaging.receiveAll] anyway
+            // Should this be the original transaction or does the sent transaction need to be the transaction signed
+            // by the previously receiving peer? If so then the multiple send and receives won't work.
+            // session.send(signedTransaction)
```
It should be ok to send everyone the initially signed tx and assemble the signatures after the fact.

blsemo commented on libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/transaction/ConsensualTransactionBuilderImpl.kt:
```diff
@@ -10,117 +12,124 @@ import net.corda.ledger.common.impl.transaction.WireTransactionDigestSettings
+    override fun signInitial(publicKey: PublicKey): ConsensualSignedTransaction {
+        val wireTransaction = buildWireTransaction()
+        // TODO(CORE-5091 we just fake the signature for now...)
+        val signature = signingService.sign(wireTransaction.id.bytes, publicKey, SignatureSpec.ECDSA_SHA256)
```
That doesn't look quite right - we should sign over bytes + metadata, no? Could probably be fixed as part of CORE-5091.

vlajos replied:
I hope that I can fix both `signInital()` and this in https://github.com/corda/corda-runtime-os/pull/2248
(Assuming the current PR gets merged first.)

chrisr3 commented on components/flow/flow-service/src/main/kotlin/net/corda/flow/application/services/MemberLookupImpl.kt:
```diff
@@ -15,7 +15,11 @@ import org.osgi.service.component.annotations.Reference
+@Component(
+    service = [MemberLookup::class, SingletonSerializeAsToken::class],
+    scope = ServiceScope.PROTOTYPE,
+    property = ["corda.system=true"]
```
Are you turning _every_ service into a "system service" :exploding_head:? A service is only supposed to be declared as a system service to allow us to `@CordaInject` it into a system flow via a non-API interface. I.e. an interface that doesn't match `net.corda.v5.*`.

lankydan replied:
If I do not include it, I get this:

```
java.lang.IllegalArgumentException: No registered types could be found for the following field(s) 'memberLookup'
	at net.corda.flow.pipeline.sandbox.impl.SandboxDependencyInjectorImpl.injectServices(SandboxDependencyInjectorImpl.kt:42) ~[?:?]
	at net.corda.flow.application.services.FlowEngineImpl.subFlow$lambda-1(FlowEngineImpl.kt:53) ~[?:?]
	at java.security.AccessController.doPrivileged(Native Method) ~[?:?]
	at net.corda.flow.application.services.FlowEngineImpl.subFlow(FlowEngineImpl.kt:52) ~[?:?]
	at net.corda.ledger.consensual.impl.ConsensualLedgerServiceImpl.finality(ConsensualLedgerServiceImpl.kt:51) ~[?:?]
	at net.cordapp.demo.consensual.ConsensualFlow.call(ConsensualFlow.kt:107) ~[?:?]
	at net.corda.flow.fiber.RPCStartedFlow.invoke(RPCStartedFlow.kt:14) ~[?:?]
	at net.corda.flow.fiber.FlowFiberImpl.runFlow(FlowFiberImpl.kt:88) ~[?:?]
	at net.corda.flow.fiber.FlowFiberImpl.run(FlowFiberImpl.kt:60) ~[?:?]
	at net.corda.flow.fiber.FlowFiberImpl.run(FlowFiberImpl.kt:20) ~[?:?]
	at co.paralleluniverse.fibers.Fiber.run1(Fiber.java:1164) ~[?:?]
	at co.paralleluniverse.fibers.Fiber.exec(Fiber.java:860) ~[?:?]
	at co.paralleluniverse.fibers.RunnableFiberTask.doExec(RunnableFiberTask.java:100) ~[?:?]
	at co.paralleluniverse.fibers.RunnableFiberTask.run(RunnableFiberTask.java:91) ~[?:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) ~[?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
	at co.paralleluniverse.concurrent.util.ScheduledSingleThreadExecutor$ScheduledFutureTask.access$001(ScheduledSingleThreadExecutor.java:160) ~[?:?]
	at co.paralleluniverse.concurrent.util.ScheduledSingleThreadExecutor$ScheduledFutureTask.run(ScheduledSingleThreadExecutor.java:272) ~[?:?]
	at co.paralleluniverse.concurrent.util.ScheduledSingleThreadExecutor.work(ScheduledSingleThreadExecutor.java:119) ~[?:?]
	at java.lang.Thread.run(Thread.java:829) ~[?:?]
```

It only works with the property, or if I reference it in one of my cordapps.

lankydan commented:
merge comment:

```
CORE-6871 Consensual Finality Flow

Initial implementation of `ConsensualFinalityFlow` and 
`ConsensualReceiveFinalityFlow`.

This includes a lot of code that is still in a TODO state, but a 
starting structure of finality has been put in place with this work.

`ConsensualTransactionBuilderFactory` has been added to split out some 
dependencies from the `ConsensualLedgerService`.

A number of services were made "system services" as they were not being 
loaded into the sandbox.

Include runtime dependencies on the serialization modules so that 
they're actually used in flows.
```

lankydan committed:
CORE-6871 Consensual Finality Flow

Initial implementation of `ConsensualFinalityFlow` and
`ConsensualReceiveFinalityFlow`.

This includes a lot of code that is still in a TODO state, but a
starting structure of finality has been put in place with this work.

`ConsensualTransactionBuilderFactory` has been added to split out some
dependencies from the `ConsensualLedgerService`.

A number of services were made "system services" as they were not being
loaded into the sandbox.

Include runtime dependencies on the serialization modules so that
they're actually used in flows.

ronanbrowne reviewed (approved):
LGTM (build)
